File: app/src/main/java/com/greenwald/aaron/ridetracker/model/Trip.java
package com.greenwald.aaron.ridetracker.model;

import com.google.android.gms.maps.model.LatLng;

import java.io.Serializable;
import java.util.ArrayList;

public class Trip implements Serializable {
    private String name;
    private long id;
    private ArrayList<Segment> segmentList = new ArrayList<>();

    public Trip(String name) {
        this.name = name;
    }

    public Trip(long id, String name) {
        this.id = id;
        this.name = name;
    }

    public String getName() {
        return name;
    }

    public long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public ArrayList<LatLng> getAllLocations() {
        ArrayList<LatLng> result = new ArrayList<>();
        //I wish I was using scala right now....need to switch to Kotlin
        for (Segment segment : segmentList)
            for (SegmentPoint segmentPoint : segment.getSegmentPoints())
                result.add(segmentPoint.getLatLng());

        return result;
    }

    public void setSegments(ArrayList<Segment> segments) {
        this.segmentList = segments;
    }

    public SegmentPoint getStartingPoint() {
        if (!segmentList.isEmpty() && !segmentList.get(0).getSegmentPoints().isEmpty()) {
            return segmentList.get(0).getSegmentPoints().get(0);
        }
        return null;
    }
}

File: app/src/main/java/com/greenwald/aaron/ridetracker/model/Trip.kt
package com.greenwald.aaron.ridetracker.model

import com.google.android.gms.maps.model.LatLng

import java.io.Serializable

data class Trip (val name: String,
                 val id: Long = 0,
                 val distance: Kilometers = Kilometers(0.0),
                 val ridingTime: Milliseconds = Milliseconds(0),
                 val elapsedTime: Milliseconds = Milliseconds(0),
                 val maxSpeed: KilometersPerHour = KilometersPerHour(0.0),
                 val maxAltitude: Meters = Meters(0.0),
                 val minAltitude: Meters = Meters(0.0),
                 val segments: List<Segment> = emptyList()) : Serializable {

    val averageRidingSpeed: KilometersPerHour get() = KilometersPerHour.from(distance, ridingTime)
    val altitudeRange: Meters get() = Meters(maxAltitude.value - minAltitude.value)
    val stoppedTime: Milliseconds get() = Milliseconds(elapsedTime.value - ridingTime.value)

    fun getAllPoints(): List<LatLng> =
            segments.flatMap { segment -> segment.segmentPoints.map(SegmentPoint::latLng) }
}
